// webex/dist/meetings.js
"use strict";

/*!
 * Copyright (c) 2015-2023 Cisco Systems, Inc. See the LICENSE file.
 */

// Note: this file is written using commonjs instead of import/export to
// simplify consumption by those less familiar with the current state of
// JavaScript modularization

/* istanbul ignore else */
if (!global._babelPolyfill) {
  /* eslint global-require: [0] */
  require('@babel/polyfill');
}
require('@webex/plugin-authorization');
// explicitly load wdm, since we're relying on preDiscoveryServices and the
// url interceptor
require('@webex/plugin-logger');
require('@webex/common');
require('@webex/plugin-meetings');
require('@webex/internal-plugin-device');
require('@webex/internal-plugin-metrics');
require('@webex/internal-plugin-support');
require('@webex/internal-plugin-user');
require('@webex/plugin-people');
var merge = require('lodash/merge');
var WebexCore = require('@webex/webex-core').default;
var config = require('./config');
var Webex = WebexCore.extend({
  webex: true,
  version: "2.59.8"
});
Webex.init = function init() {
  var attrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  attrs.config = merge({
    sdkType: 'meetings',
    meetings: {
      disableHydraId: true
    }
  }, config, attrs.config); // eslint-disable-line no-param-reassign

  return new Webex(attrs);
};
module.exports = Webex;
