// @webex/plugin-meetings/dist/reactions/reactions.type.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.SkinToneType = exports.ReactionType = void 0;
// @ts-ignore
var ReactionType;
exports.ReactionType = ReactionType;
(function (ReactionType) {
  ReactionType["smile"] = "smile";
  ReactionType["sad"] = "sad";
  ReactionType["wow"] = "wow";
  ReactionType["haha"] = "haha";
  ReactionType["celebrate"] = "celebrate";
  ReactionType["clap"] = "clap";
  ReactionType["thumbs_up"] = "thumbs_up";
  ReactionType["thumbs_down"] = "thumbs_down";
  ReactionType["heart"] = "heart";
  ReactionType["fire"] = "fire";
  ReactionType["prayer"] = "prayer";
  ReactionType["speed_up"] = "speed_up";
  ReactionType["slow_down"] = "slow_down";
})(ReactionType || (exports.ReactionType = ReactionType = {}));
var SkinToneType;
exports.SkinToneType = SkinToneType;
(function (SkinToneType) {
  SkinToneType["normal"] = "normal";
  SkinToneType["light"] = "light";
  SkinToneType["medium_light"] = "medium_light";
  SkinToneType["medium"] = "medium";
  SkinToneType["medium_dark"] = "medium_dark";
  SkinToneType["dark"] = "dark";
})(SkinToneType || (exports.SkinToneType = SkinToneType = {}));
